build_tools/artifact_manager.py

```diff
@@ -0,0 +1,787 @@
+directories (for prototyping) and S3 (for CI/CD). Operations are parallelized
+for performance.
+
+Usage:
+    # Fetch inbound artifacts for a stage (downloads and extracts in parallel)
+    python stage_artifact_manager.py fetch \
+        --stage math-libs \
+        --amdgpu-families gfx94X-dcgpu \
+        --run-id 12345 \
+        --output-dir build/
+
+    # Push produced artifacts after building (compresses and uploads in parallel)
+    python stage_artifact_manager.py push \
+        --stage math-libs \
+        --amdgpu-families gfx94X-dcgpu \
+        --run-id 12345 \
+        --build-dir build/
+
+    # List what artifacts a stage needs/produces
+    python stage_artifact_manager.py info \
+        --stage math-libs \
+        --amdgpu-families gfx94X-dcgpu
+
+Environment variables:
+    THEROCK_LOCAL_STAGING_DIR: Use local directory instead of S3
+    THEROCK_RUN_ID: Override run ID
+    THEROCK_PLATFORM: Override platform (default: linux)
+"""
+
+import argparse
+import concurrent.futures
+from dataclasses import dataclass
+import os
+import shutil
+import sys
+import tarfile
+import threading
+import time
+from pathlib import Path
+from typing import List, Optional, Set
+
+# Add build_tools to path for imports
+sys.path.insert(0, str(Path(__file__).parent))
+
+from _therock_utils.build_topology import BuildTopology
+from _therock_utils.artifact_backend import (
+    ArtifactBackend,
+    create_backend_from_env,
+)
+from _therock_utils.artifacts import ArtifactName, ArtifactPopulator
+
+
+def log(msg: str):
+    """Print message and flush."""
+    print(msg, flush=True)
+
+
+def get_topology() -> BuildTopology:
+    """Load the BUILD_TOPOLOGY.toml from the repository root."""
+    script_dir = Path(__file__).parent
+    repo_root = script_dir.parent
+    topology_path = repo_root / "BUILD_TOPOLOGY.toml"
+    if not topology_path.exists():
+        raise FileNotFoundError(f"BUILD_TOPOLOGY.toml not found at {topology_path}")
+    return BuildTopology(str(topology_path))
+
+
+# =============================================================================
+# Fetch (Download + Extract) with Parallel Processing
+# =============================================================================
+
+
+@dataclass
+class DownloadRequest:
+    """Request to download an artifact."""
+
+    artifact_key: str
+    dest_path: Path
+    backend: ArtifactBackend
+
+
+@dataclass
+class ExtractRequest:
+    """Request to extract a downloaded artifact."""
+
+    archive_path: Path
+    output_dir: Path
+    delete_archive: bool
+    flatten: bool
+    bootstrap: bool = False
+    # Shared state for parallel bootstrap extraction
+    cleaned_paths: Optional[set] = None
+    cleaned_paths_lock: Optional[threading.Lock] = None
+
+
+def download_artifact(request: DownloadRequest) -> Optional[Path]:
+    """Download a single artifact with retry logic."""
+    MAX_RETRIES = 3
+    BASE_DELAY = 2
```

Comment:
nit: Let's add units to `BASE_DELAY`, like `BASE_DELAY_SECONDS` or `BASE_DELAY_MS`

`time.sleep()` accepts `secs`: https://docs.python.org/3/library/time.html#time.sleep

Ah, this code is forked from https://github.com/ROCm/TheRock/commit/d3809fe83f6d931d02884f94e81b193f1e2321d1, which I approved x_x